cql3/statements/select_statement.cc

```diff
@@ -461,8 +467,12 @@ select_statement::execute_without_checking_exception_message_aggregate_or_paged(
             state, options, command, std::move(key_ranges), _restrictions_need_filtering ? _restrictions : nullptr);
 
     if (aggregate || nonpaged_filtering) {
-        auto builder = cql3::selection::result_set_builder(*_selection, now, *_group_by_cell_indices);
-        coordinator_result<void> result_void = co_await utils::result_do_until([&p] {return p->is_exhausted();},
+        uint64_t limit = _limit ? get_limit_impl(options, *_limit) : std::numeric_limits<uint64_t>::max();
+        auto builder = cql3::selection::result_set_builder(*_selection, now, *_group_by_cell_indices, limit);
```

Comment:
You still haven't explained why we need to evaluate the limit again. I see it's already evaluated in select_statement::do_execute(). 

I'm also confused (independently of the patch) about the too many places the limit goes into. select_statement::do_execute() embeds the limit into read_command. But if we're aggregating, we need to apply the limit to the aggregated result, not to the input rows.

Reply:
Wait I know why. get_limit() doesn't just get the limit, it does something else if is_aggregate.

Reply:
Here's a better solution:
1. break out the extra checks in get_limit() and add them to the callers.
2. Refine the extra checks at the right callers. For example, we can have two variables: inner_loop_limit and outer_loop_limit. For aggregation, we'd set outer_loop_limit to the LIMIT, otherwise inner_loop_limit.

Why it's better: instead of having multiple layers of get_limit() functions with no explanation, and a get_limit() that is not simple to understand (because it does another calculation), we have a trivial get_limit() and each caller gets to choose what it does with it.

Reply:
get_limit_with_default_and_unset_guard() still has `|| is_aggregate()`. So it's returning the wrong answer and expecting the caller to know which get_limit() variant to call.

The is_aggregate() call should move to the call sites.

See below (not even compile tested)

```diff
diff --git a/cql3/statements/select_statement.cc b/cql3/statements/select_statement.cc
index 2e29bac1de..1295f591d6 100644
--- a/cql3/statements/select_statement.cc
+++ b/cql3/statements/select_statement.cc
@@ -356,11 +356,18 @@ select_statement::do_execute(query_processor& qp,
 
     auto cl = options.get_consistency();
 
     validate_for_read(cl);
 
-    uint64_t limit = get_limit(options);
+    uint64_t inner_loop_limit = get_limit(options);
+    uint64_t outer_loop_limit = inner_loop_limit; // If we don't aggregate, the outer loop and the inner loop are the same loop
+
+    if (is_aggregate() || needs_post_query_ordering() /* ?? */) {
+        // If we have an outer loop, LIMIT applies to the outer loop.
+        inner_loop_limit = std::numeric_limits<uint64_t>::max();
+    }
+
     auto now = gc_clock::now();
 
     _stats.filtered_reads += _restrictions_need_filtering;
 
     const source_selector src_sel = state.get_client_state().is_internal()
@@ -378,11 +385,11 @@ select_statement::do_execute(query_processor& qp,
             _schema->id(),
             _schema->version(),
             std::move(slice),
             max_result_size,
             query::tombstone_limit(qp.proxy().get_tombstone_limit()),
-            query::row_limit(limit),
+            query::row_limit(inner_loop_limit),
             query::partition_limit(query::max_partitions),
             now,
             tracing::make_trace_info(state.get_trace_state()),
             query_id::create_null_id(),
             query::is_first_page::no,
@@ -436,11 +443,11 @@ select_statement::do_execute(query_processor& qp,
     if (!aggregate && !_restrictions_need_filtering && (page_size <= 0
             || !service::pager::query_pagers::may_need_paging(*_schema, page_size,
                     *command, key_ranges))) {
         f = execute_without_checking_exception_message_non_aggregate_unpaged(qp, command, std::move(key_ranges), state, options, now);
     } else {
-        f = execute_without_checking_exception_message_aggregate_or_paged(qp, command, std::move(key_ranges), state, options, now, page_size, aggregate, nonpaged_filtering);
+        f = execute_without_checking_exception_message_aggregate_or_paged(qp, command, std::move(key_ranges), state, options, now, page_size, aggregate, nonpaged_filtering, outer_loop_limit);
     }
 
     if (!tablet_info.has_value()) {
         return f;
     }
@@ -452,19 +459,19 @@ select_statement::do_execute(query_processor& qp,
 }
 
 future<::shared_ptr<cql_transport::messages::result_message>>
 select_statement::execute_without_checking_exception_message_aggregate_or_paged(query_processor& qp,
         lw_shared_ptr<query::read_command> command, dht::partition_range_vector&& key_ranges, service::query_state& state,
-        const query_options& options, gc_clock::time_point now, int32_t page_size, bool aggregate, bool nonpaged_filtering) const {
+        const query_options& options, gc_clock::time_point now, int32_t page_size, bool aggregate, bool nonpaged_filtering, uint64_t outer_loop_limit) const {
     command->slice.options.set<query::partition_slice::option::allow_short_read>();
     auto timeout_duration = get_timeout(state.get_client_state(), options);
     auto timeout = db::timeout_clock::now() + timeout_duration;
     auto p = service::pager::query_pagers::pager(qp.proxy(), _schema, _selection,
             state, options, command, std::move(key_ranges), _restrictions_need_filtering ? _restrictions : nullptr);
 
     if (aggregate || nonpaged_filtering) {
-        auto builder = cql3::selection::result_set_builder(*_selection, now, *_group_by_cell_indices);
+        auto builder = cql3::selection::result_set_builder(*_selection, now, *_group_by_cell_indices, outer_loop_limit);
         coordinator_result<void> result_void = co_await utils::result_do_until([&p] {return p->is_exhausted();},
                 [&p, &builder, page_size, now, timeout] {
                     return p->fetch_page_result(builder, page_size, now, timeout);
                 }
         );
```